### PR TITLE
680 - accept/reject: Allow users to submit nodes and resubmit rejected nodes

### LINKED ISSUE #680: Allow users to submit nodes and resubmit rejected nodes
Instead of the submit buttons, user would see:
    "Save as Private Draft"
    "Submit to Main Tapestry"

If the node has already been rejected, instead of "Submit", it should say "Resubmit"

### PR DESCRIPTION
**Pull request by liam-armstrong**
Closes #680 
Adds:
 - Node visibility controls based on status attribute:
Nodes with `publish` can be seen by all users
Nodes with `draft` and `reject` can only be seen by their author
Nodes with `submitted` can only be seen by their author and Tapestry administrators
- Editors see a review form when editing `submitted` nodes, and are given the option to publish or reject a node, with a comment
- Comments are stored as an array of objects related to each node. Each comment stores the rich text comment, a time stamp in ISO 8601 format, and comment author's name and email
- Nodes on the frontend have different border colors based on status. The colors currently here are just placeholders, but demonstrate what can be done.

### CODE REVIEW

**Comment by seanRong** [comment]
After testing, I think everything here works ok with regards to publish/draft and reject/accept. I might have to test again after fixing the issue in 674 though.

**Comment by seanRong** [comment]
Verified that the non-reviewer version of the modal is the default, so under laggy circumstances the reviewer is the one who will briefly see the submitter modal, which shouldn't be an issue.

**Review by narendrasss — APPROVED** [review]
Code LGTM, once tested should be good to go.

**Comment by seanRong** [comment]
Submission form issues should be fixed here

**Comment by jaskwong** [comment]
- After submitted nodes are approved and published, the author of the node can still add/edit/delete the node (and save it as a private draft again)
- The administrator cannot change the permissions for the author (but these 2 are like the same problem)

**Comment by seanRong** [comment]
Merge confs will be addressed once 674 fully done

**Comment by jaskwong** [comment]
Tested, works great!

**Comment by jaskwong** [comment]
Creating a node and adding 2 links to the node (like a triangle) and then submitting it, if the admin rejects it, the user can no longer resubmit the node nor save as private draft

**Comment by seanRong** [comment]
@jaskwong This shouldn't even be possible, irrc. How did you manage to add more than 1 link to a draft node?

**Comment by jaskwong** [comment]
@seanRong it's a pretty obscure case but if the user has add privileges to another node, then I was able to add a link from that node to the new node

**Comment by seanRong** [comment]
NOTE: There are some issues even after resolving conflicts

**Comment by narendrasss** [comment]
Assigning this to myself to add integration tests.

**Comment by narendrasss** [comment]
@seanRong @wynnset I notice that when a node is accepted, the author of the node is set to the _reviewer_ rather than the original node author. Is this intended? Would it not be more appropriate to leave the author as the original submitter?

**Comment by wynnset** [comment]
> @seanRong @wynnset I notice that when a node is accepted, the author of the node is set to the _reviewer_ rather than the original node author. Is this intended? Would it not be more appropriate to leave the author as the original submitter?

Yes, the author should not be changed to the reviewer. It should remain as the original author.

**Comment by seanRong** [comment]
@wynnset Looking back in the commit history there's one that explicitly changes the author from the submitter to the admin. Do you remember why this was requested?

**Comment by wynnset** [comment]
@seanRong just looking at the comment history here and found [this](https://github.com/wynnset/tapestry-wp/pull/683#issuecomment-702421676) by @jaskwong:
> * After submitted nodes are approved and published, the author of the node can still add/edit/delete the node (and save it as a private draft again)
> * The administrator cannot change the permissions for the author (but these 2 are like the same problem)

This makes sense in that the person who submitted the node should not be able to edit a submitted node. However, we do want to have them remain as the author for attribution / copyright purposes. We should keep the author the same, but not allow them to make changes anymore after it's submitted. We can do this by either:
1. Marking the accepted nodes with a different status (e.g. "accepted") rather than just "publish", OR
2. Create a new property to denote a particular node was a submitted node originally.
We can then ensure we do not allow the author of the node to add/edit any nodes with such a distinction.

Does that make sense? Open to other ideas.

**Comment by seanRong** [comment]
PR 712 actually has the accepted status versus just the normal published, so it could be differentiated during that PR. Would it be fine to leave it for now?

**Comment by narendrasss** [comment]
@seanRong gonna assign this back to you to discuss Aidin's comment above. I added some tests for the `ReviewForm` and fixed a minor edge case. There's more tests to add here but I think they belong under e2e/Cypress.

**Comment by wynnset** [comment]
As discussed, please add the feature I mentioned above into this PR. Also, as discussed, the scope of this PR can be expanded to include the actual accept/reject part of the workflow by the reviewer and PR #712 can be changed to be only for the commenting part. Thanks @seanRong

**Comment by jaskwong** [comment]
Tested, works well!

**Comment by wynnset** [comment]
If I submit a node, it should not allow me to edit it anymore (it currently does). This task requires the approval process to happen within the sidebar in a new "Review" section (under the copyright section), with the review comments in there. I currently see cannot see the approval process in the sidebar. Instead I have to press "Edit" to be able to approve/reject something. Additionally, I cannot see the comment thread either. Also, the "Accept" button is supposed to read "Accept and Add".

**Comment by seanRong** [comment]
> This task requires the approval process to happen within the sidebar in a new "Review" section (under the copyright section), with the review comments in there. I currently see cannot see the approval process in the sidebar. Instead I have to press "Edit" to be able to approve/reject something. Additionally, I cannot see the comment thread either.

These changes were made in 712, and weren't part of the original scope for 680. (Asana reflects that)

**Comment by wynnset** [comment]
> These changes were made in 712, and weren't part of the original scope for 680. (Asana reflects that)

Ok we can move the "Accept / Reject" buttons to the sidebar in #712, then, and also do the commenting part in there (as planned).

File: templates/vue/src/components/NodeModal.vue
Comment (wynnset) [inline templates/vue/src/components/NodeModal.vue:308]: If a node is published, why would it matter what the status of its neighbours are?

File: templates/vue/src/components/NodeModal.vue
Comment (wynnset) [inline templates/vue/src/components/NodeModal.vue:329]: All methods and calculated values should start with a verb for clarity. This should read `canEditTapestry`.

File: templates/vue/src/components/NodeModal.vue
Comment (wynnset) [inline templates/vue/src/components/NodeModal.vue:558]: It's not easy to understand why these conditions are here. It would be great if you could create some calculated values that are named in a way to describe what you are checking for exactly or otherwise add a comment to say why this condition is being added.

File: templates/vue/src/components/node-modal/DeleteNodeButton.vue
Comment (wynnset) [inline templates/vue/src/components/node-modal/DeleteNodeButton.vue:40]: Nodes cannot be made into drafts.

File: templates/vue/src/utils/Helpers.js
Comment (wynnset) [inline templates/vue/src/utils/Helpers.js:151]: These conditions are very confusing. Why are accepted and submitted nodes not permitted, but the other ones are? Can you explain that in a comment to clarify or otherwise think how you can put this code in a way that is easier to understand?

**Review by wynnset — CHANGES_REQUESTED** [review]
Sorry @seanRong, this still needs a bit more work to become functional and maintainable. I've made some comments in the code but also:
- Accepted nodes should have a progress bar same as published nodes.
- Accepted nodes should appear for everyone. They are the same as published nodes. Right now, they only appear to the author of the node and admins.

I find adding of all the conditions around statuses has made the code a lot harder to read and understand. Maybe it's better if we create a different field to designate a node as accepted rather than using the status field - that way we won't have to check everywhere in the code for both. We might also want to do the same with other fields so that we only have draft and published as statuses. We could have a new field called `reviewStatus` and have it set to either submitted, rejected, accepted, or null (in case it was never reviewed).

File: templates/vue/src/components/NodeModal.vue
Comment (seanRong) [reply templates/vue/src/components/NodeModal.vue:329]: Done

File: templates/vue/src/components/NodeModal.vue
Comment (seanRong) [reply templates/vue/src/components/NodeModal.vue:125]: Added

File: templates/vue/src/components/node-modal/DeleteNodeButton.vue
Comment (seanRong) [reply templates/vue/src/components/node-modal/DeleteNodeButton.vue]: Probably missed this in merge, fixed.

File: classes/class.tapestry-node.php
Comment (seanRong) [reply classes/class.tapestry-node.php:41]: Changed

File: templates/vue/src/components/NodeModal.vue
Comment (seanRong) [reply templates/vue/src/components/NodeModal.vue]: Gave me a headache until I remembered <template> existed haha

File: templates/vue/src/components/NodeModal.vue
Comment (seanRong) [reply templates/vue/src/components/NodeModal.vue:310]: Yeah this check doesn't make much sense - removed.

**Comment by seanRong** [comment]
> Accepted nodes should have a progress bar same as published nodes.
Accepted nodes should appear for everyone. They are the same as published nodes. Right now, they only appear to the author of the node and admins.

These  two points have now been addressed.

File: templates/vue/src/components/NodeModal.vue
Comment (seanRong) [reply templates/vue/src/components/NodeModal.vue]: Clarified

File: templates/vue/src/utils/Helpers.js
Comment (seanRong) [reply templates/vue/src/utils/Helpers.js]: Clarified with minor change

File: templates/vue/src/components/NodeModal.vue
Comment (wynnset) [inline templates/vue/src/components/NodeModal.vue:271]: I had said [before](https://github.com/wynnset/tapestry-wp/pull/683#discussion_r516274856) (I think it might have been missed in between my other feedback): The concept of `this.parent` does not exist within edit context. How do you decide which node is the parent if the node is connected to more than a single other node? That is why the `else if` statement was added. I think you'd need to revert so that we check ALL the neighbouring nodes and see if ANY of them are not published or accepted.

File: templates/vue/src/components/NodeModal.vue
Comment (wynnset) [inline templates/vue/src/components/NodeModal.vue:321]: Please use the new `wp` service to be consistent.

File: templates/vue/src/components/NodeModal.vue
Comment (wynnset) [inline templates/vue/src/components/NodeModal.vue:503]: Just checking.. was this removed on purpose? I don't see it in any other part so wanted to make sure it's not a mistake.

File: templates/vue/src/components/NodeModal.vue
Comment (wynnset) [inline templates/vue/src/components/NodeModal.vue:924]: I suggest you use [bootstrap margin classes](https://getbootstrap.com/docs/4.3/utilities/spacing/) instead of adding a custom style.

File: templates/vue/src/components/tapestry-node/ProgressBar.vue
Comment (wynnset) [inline templates/vue/src/components/tapestry-node/ProgressBar.vue:40]: I don't think this is needed anymore.

File: templates/vue/src/components/TapestryNode.vue
Comment (wynnset) [inline templates/vue/src/components/TapestryNode.vue:37]: Could you please name this something more meaningful like `status-bar` maybe? Also, submitted, accepted, and rejected nodes should not be dashed - they should be solid. Only simple draft nodes should have a dashed line. Please also make it one or two pixels thicker. It's a bit hard to see when the nodes are small.

File: templates/vue/src/components/node-modal/ReviewForm.vue
Comment (wynnset) [inline templates/vue/src/components/node-modal/ReviewForm.vue:49]: This line could be safely removed, I believe.

File: templates/vue/src/utils/Helpers.js
Comment (wynnset) [inline templates/vue/src/utils/Helpers.js:151]: You are erroneously removing the ability for node authors to edit their own node here (if their node is not draft). Published nodes should remain editable by their authors (unless they were submitted and accepted).

File: templates/vue/src/utils/constants.js
Comment (wynnset) [inline templates/vue/src/utils/constants.js:40]: a) this is never used anywhere
b) if you are planning to use this, please split it into `nodeStatuses` and `reviewStatuses`

**Review by wynnset — CHANGES_REQUESTED** [review]
The updates look really good. Just a few small fixes and we are golden!

File: templates/vue/src/components/tapestry-node/ProgressBar.vue
Comment (seanRong) [reply templates/vue/src/components/tapestry-node/ProgressBar.vue]: Yeah, this prop is redundant.

File: templates/vue/src/utils/constants.js
Comment (seanRong) [reply templates/vue/src/utils/constants.js]: Deleted (might be a good idea in the future though?)

File: templates/vue/src/components/node-modal/ReviewForm.vue
Comment (seanRong) [reply templates/vue/src/components/node-modal/ReviewForm.vue]: Yep.

File: templates/vue/src/components/NodeModal.vue
Comment (seanRong) [reply templates/vue/src/components/NodeModal.vue:924]: OK this one, I don't really want to change, as I would need to put a margin class on every single button under that div. Thoughts?

Also I don't think 0.25rem is one of the bootstrap defaults

File: templates/vue/src/components/NodeModal.vue
Comment (seanRong) [reply templates/vue/src/components/NodeModal.vue:503]: This was removed because it is almost an exact duplicate of the error checking under handleSubmit(). It probably does not turn up if you cmd+F for it, because there's minute differences that set node status to draft in there.

File: templates/vue/src/components/NodeModal.vue
Comment (wynnset) [reply templates/vue/src/components/NodeModal.vue:271]: Why was this marked as resolved? It has not been resolved...

Comment (seanRong) [reply templates/vue/src/components/NodeModal.vue]: Sorry, I didn't actually push!

**Comment by narendrasss** [comment]
@seanRong I pushed a super small commit that hopefully fixes the cy tests

**Review by wynnset — APPROVED** [review]

File: templates/vue/src/components/NodeModal.vue
Comment (narendrasss) [inline templates/vue/src/components/NodeModal.vue:526]: @wynnset Any reason why this was changed to not use async await? I think it's cleaner the way it was before.

**Review by narendrasss — CHANGES_REQUESTED** [review]

File: templates/vue/src/components/NodeModal.vue
Comment (wynnset) [reply templates/vue/src/components/NodeModal.vue:526]: I changed this because the `loading` data property was immediately changing to false and so the node modal was not disabling the buttons while that was happening and I could click cancel or submit again while the node was submitting. Do you have a suggestion of how to avoid that in a different way?

**Review by narendrasss — APPROVED** [review]

**Pull request merged**
